cmd/router/main.go

```diff
@@ -351,11 +365,43 @@ func main() {
 		os.Exit(1)
 	}
 
+	server := &http.Server{
+		Addr:    ":" + strconv.Itoa(constants.RouterPort),
+		Handler: nil, // default server mux
+		// https://medium.com/a-journey-with-go/go-understand-and-mitigate-slowloris-attack-711c1b1403f6
+		ReadHeaderTimeout: time.Minute,
+	}
 	http.HandleFunc("/", graphHandler)
+	http.HandleFunc(constants.RouterReadinessEndpoint, readyHandler)
 
-	err = http.ListenAndServe(":8080", nil)
-	if err != nil {
-		log.Error(err, "failed to listen on 8080")
+	go func() {
+		err = server.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.Error(err, fmt.Sprintf("Failed to serve on address %v", server.Addr))
+			os.Exit(1)
+		}
+	}()
+
+	// Blocks until SIGTERM or SIGINT is received
+	handleSignals(server)
+}
+
+func handleSignals(server *http.Server) {
+	signalChan := make(chan os.Signal, 1)
+	signal.Notify(signalChan, os.Interrupt, syscall.SIGTERM)
+
+	sig := <-signalChan
+	log.Info("Received shutdown signal", "signal", sig)
+	// Fail the readiness probe
+	isShuttingDown = true
+	log.Info("Sleeping %v to allow K8s propagation of non-ready state", drainSleepDuration)
+	// Sleep to give networking a little bit more time to remove the pod
+	// from its configuration and propagate that to all loadbalancers and nodes.
+	time.Sleep(drainSleepDuration)
```

Comment:
@rachitchauhan43 / @yuzisun I have added readiness probe for the router. The readiness probe will start failing after receiving the shutdown signal. The router will wait for 30 seconds before shutting down the server to accommodate the scenario mentioned in the issue #3375. It is worth noting that knative serving already adds a preStop hook to the container which essentially delays the termination to process the accepted requests. https://github.com/knative/serving/blob/51b0337fa29d413302b75acfa13bbd8c765e8fea/pkg/queue/constants.go#L24-L29
![Screenshot from 2024-01-27 23-12-25](https://github.com/kserve/kserve/assets/90966311/4ba2f299-6b2c-43a3-94d4-5a5154480f8d)
@rachitchauhan43 Would you mind testing it with these changes ?